tests/spark/upgrade/utils.py

```
@@ -0,0 +1,351 @@
+                LOGGER.info(f"SparkApplication {spark_app.name} reached state {expected_state}")
+                return
+    except TimeoutExpiredError:
+        status = spark_app.instance.status
+        if status is None:
+            current_state = "No status yet"
+        else:
+            current_state = status.get("applicationState", {}).get("state", "UNKNOWN")
+        raise TimeoutExpiredError(
+            f"SparkApplication {spark_app.name} did not reach {expected_state} state within {timeout}s. "
+            f"Current state: {current_state}"
+        )
```

Ruff (0.15.17)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)